source/common/stats/symbol_table_impl.cc

```diff
@@ -117,76 +144,124 @@ SymbolEncoding SymbolTable::encode(const absl::string_view name) {
 }
 
 std::string SymbolTable::decode(const SymbolStorage symbol_array, uint64_t size) const {
-  return decode(SymbolEncoding::decodeSymbols(symbol_array, size));
+  return decodeSymbolVec(SymbolEncoding::decodeSymbols(symbol_array, size));
 }
 
-std::string SymbolTable::decode(const SymbolVec& symbols) const {
+std::string SymbolTable::decodeSymbolVec(const SymbolVec& symbols) const {
   std::vector<absl::string_view> name_tokens;
   name_tokens.reserve(symbols.size());
   {
     // Hold the lock only while decoding symbols.
-    Thread::LockGuard lock(lock_);
+    absl::ReaderMutexLock lock(&lock_);
     for (Symbol symbol : symbols) {
       name_tokens.push_back(fromSymbol(symbol));
     }
   }
   return absl::StrJoin(name_tokens, ".");
 }
 
-void SymbolTable::free(StatName stat_name) {
+void SymbolTable::incRefCount(const StatName& stat_name) {
   // Before taking the lock, decode the array of symbols from the SymbolStorage.
-  SymbolVec symbols = SymbolEncoding::decodeSymbols(stat_name.data(), stat_name.numBytes());
+  SymbolVec symbols = SymbolEncoding::decodeSymbols(stat_name.data(), stat_name.dataSize());
 
-  Thread::LockGuard lock(lock_);
+  absl::ReaderMutexLock lock(&lock_);
```

Comment:
From a quick read and without entirely paging everything back in, it seems odd to me that we are incrementing some reference count but only taking a read lock. I'm probably missing something but are there any more comments or a threading overview somewhere that would help here?

Reply:
Yes I think I commented elsewhere that with the ref-count as an atomic, an increment is safe to do under reader-lock. A decrement is not, though, as if it goes to zero we will mutate the map.

I'll add more comments.